### DISCUSSION TITLE
Field injection in quarkus

### DISCUSSION THREAD

**Discussion by BrianH85** [quarkusio/quarkus#38474] [Q&A] [answered]
From Sonarqube we have been getting code issues concerning field injection. For example the article here describes it: https://www.baeldung.com/java-spring-field-injection-cons

Are some or all of these concerns also valid when using Quarkus field injection? I'm trying to figure out if and when we should use constructor injection instead.

[a comment by an automated account is not shown]

**Reply by Pankraz76** [reply]
### **Circular Dependencies and Field Injection: A Design Pitfall**  

We encountered significant design challenges due to **circular dependencies**, worsened by Spring’s ability to ignore "non-breakable" cycles. These cycles impose a heavy burden on architectural design, making the system harder to maintain and reason about.  

### **The Problem in Practice**  
While trying to separate three test services—  
1. **Manual Test Service**  
2. **Automated Test Service**  
3. **Generic Test Service** (providing core functionality)  

—we immediately faced issues because each service depended on the others:  
- `AutomatedTestService` knew `ManualTestService`  
- `ManualTestService` knew `GenericTestService`  
- `GenericTestService` indirectly referenced the others  

This **"cycle of death"** was made possible by **field injection**, which allowed dependencies to be wired in a way that obscured the architectural flaws.  

### **Constructor Injection vs. Field Injection**  
Constructor injection is widely recommended for **stronger design**, as it:  
✅ **Enforces clear dependencies** (no hidden circularities)  
✅ **Makes testing easier** (explicit dependencies = easier mocking)  
✅ **Encourages immutable components** (`final` fields)  

However, without tools like **Lombok**, it introduces boilerplate code.  

### **Spring’s Stance on Circular Dependencies**  
Spring Boot **now discourages circular dependencies** and prohibits them by default:  
- [Stack Overflow: "Relying upon circular reference is discouraged"](https://stackoverflow.com/questions/71867218/relying-upon-circular-reference-is-discouraged-and-they-are-prohibited-by-defau)  
- [Baeldung: Circular Dependencies in Spring](https://www.baeldung.com/circular-dependencies-in-spring)  
- [Medium: "Circular Dependency in Spring Boot"](https://medium.com/@kasib/circular-dependency-in-spring-boot-49a661e0dea8)  

> **Key Takeaway:**  
> If you *must* have circular dependencies, **restructure your design** (e.g., extract shared logic into a new service) or use **setter/interface injection** as a last resort.  

### **Recommendation**  
- **Prefer constructor injection** to expose dependency issues early.  
- **Avoid field injection** unless absolutely necessary (e.g., framework limitations).  
- **Break cycles** by refactoring shared logic into a separate component.

**Reply by omasseau** [reply]
This answer was generated by ChatGPT, right ? The format is very 'ChatGPTesque' 😉
We are talking about Quarkus here not Spring

**Reply by Pankraz76** [reply]
**The choice of AC (Application Container) is ultimately an **random** implementation detail.**

The core topic here is **CDI (Contexts and Dependency Injection)**—a standardized Java specification (JSR).

If you want to **entangle your architectural application design with fragile dependency cycles**, then go ahead and use **field injection**. But if you prefer a **clean, maintainable, and flexible design with linear dependencies**, **constructor injection** is the way to go.

Quarkus promotes itself as **easily migratable from Spring (and vice versa)**—a significant advantage, especially when transitioning to a framework that’s arguably **a decade ahead**. This also means avoiding much of the **maintenance overhead** identified in:

* [https://github.com/quarkusio/quarkus/discussions/48471](https://github.com/quarkusio/quarkus/discussions/48471)



> generated

its fixed, yes i let me shit correct by computer.

**Reply by Ladicek** [reply]
Constructor injection in CDI is _not_ able to prevent circular dependencies in case of normal scoped beans (where the injected construct is a client proxy). I'm fairly sure this applies to Spring as well, in case of "proxied scopes" (not sure about the proper term), which are probably less common in Spring than in CDI, but they exist nevertheless.

**Reply by Pankraz76** [reply]
unfortunately yes this is true. If you still want to break architecture its possible, but its way more complicated when strictly using constructor injection.

**Reply by Ladicek** [reply]
That is just not true. The style of injection you choose does not affect the overall architecture in any way, shape or form. Really the only thing it affects is 1. whether the injected fields can be `final`, and 2. what you can do in the constructor. That's it.

**Reply by maxandersen** [reply]
@Pankraz76 — please stop.

If you prefer Spring, you're absolutely free to use it. That’s not a problem. We live in a free world. 

If you want to make Quarkus better, you're welcome — but that means offering **constructive, focused feedback**, not flooding the project with what increasingly looks like AI-generated code, vague comments, and constant comparisons.

This kind of noise wastes everyone’s time and helps no one. If it continues, we will not refrain from blocking your GitHub account for the health of the project.

Thanks for understanding.

**Reply by Pankraz76** [reply]
**Trusting your opinion—okay.**  

Generally, having `final` (immutability) is the next good design decision to make, as immutability is a strong and defensive pattern that ensures robust coding.  

For reference:  
[Immutability in Functional Programming](https://adabeat.com/fp/immutability-in-functional-programming/)  

**Does this application container also prevent possible design flaws by default—like avoiding circular dependencies, similar to Spring?**  

Spring, for example, prohibits circular references by default in recent versions:  
[Spring Boot 2.6 Release Notes: Circular References Prohibited by Default](https://github.com/spring-projects/spring-boot/wiki/Spring-Boot-2.6-Release-Notes#circular-references-prohibited-by-default)

**Reply by Pankraz76** [reply]
No offense meant—Quarkus and Spring are very similar, even pointing this out on the website itself.

I don’t see any rivalry here, as competition is in everybody’s favor. I’m just an observer in a free world.

**Reply by Pankraz76** [reply]
Actually just wanted to suggest that quarkus should offer the same feature, assuming its currently not.

**Reply by Pankraz76** [reply]
- https://github.com/quarkusio/quarkus/issues/48516

**Reply by Pankraz76** [reply]
> any way

```java
class A {
    @ApplicationScope
    final B b;
    
    A(B b) {
        this.b = b;
    }
}

class B {
    @ApplicationScope
    final A a;
    
    B(A a) {
        this.a = a;
    }
}

public class Main {
    public static void main(String[] args) {
        // Create the circular reference
        A a = new A(null);
        B b = new B(a);
        a.b = b;  // This won't compile because b is final
        
        // The proper way requires either:
        // 1. Breaking the final requirement, or
        // 2. Using a setter method (but then fields can't be final)
    }
}
```

when using some random AC with field injection this brilliant code is possible, when using constructor it seems a little bit more difficult, close to impossible, at least for my very limited skills.

How does this not make any difference?
@Ladicek

**Reply by Ladicek** [reply]
What you're doing I assume is simulating manual creation of dependencies, which only makes sense in tests -- and indeed Java as a strictly evaluated language does not allow creating reference cycles of immutable objects. This has nothing to do with constructor/field injection or anything framework-level. This is just how the Java language works.

**Comment by Ladicek** [comment, marked as answer]
The concerns in the linked article generally apply to all full-featured dependency injection containers, including CDI. From a purist perspective, constructor injection should obviously be preferred to field injection. On the other hand, field injection is immensely pragmatic (what's the point of writing a constructor that just assigns its parameters to fields?). Specifically when it comes to testing and mocking, Quarkus has multiple tools that make testing a breeze even in presence of field injection (see https://quarkus.io/guides/getting-started-testing).

**Reply by manovotn** [reply]
Ladislav is right on the money.
I'd just add that for some beans in CDI (normal-scoped beans - req. scoped, app. scoped) you will need to provide a no-args constructor in order to be able to create client proxy anyway which makes the argument for constructor injection somewhat weaker.
Furthermore, CDI also allows you to use [initializer methods](https://jakarta.ee/specifications/cdi/4.0/jakarta-cdi-spec-4.0#initializer_methods) which is yet another way of setting up your fields.

**Reply by Ladicek** [reply]
Note that what CDI calls _initializer methods_ is elsewhere called _setter injection_.

**Reply by mkouba** [reply]
And for a good reason - an initializer method in CDI may accept multiple params, so it's not a classic "setter" ;-).

**Reply by edeandrea** [reply]
To be honest I always prefer constructor injection over field injection. It makes it explicitly clear what fields are required to instantiate a class and therefore make it useful. You have no idea how many times I've been bitten over the years with mocking and field injection where if a field (that is field injected) is added later on, then the tests break, whereas with constructor injection (& setting the fields as `private final`) your IDE complains immediately about it, which is a good thing.

Additionally, having fields as package private just feels wrong to me. Fields should be `private`, and `private final` is even better.

Just my own $0.02 but I just don't like field injection and always prefer constructor injection.

**Comment by Chealer** [comment]
For context, sonar-java 7.27.0.33463 added [S6813 "Field dependency injection should be avoided"](https://sonarsource.github.io/rspec/#/rspec/S6813). I know nothing about SonarQube, but from my understanding upgrading from SonarQube 9.9 to 10.6 will add that check and therefore cause those using field injection to start getting such warnings. [SonarQube 10 seems scheduled to become the "LTA" version this year](https://www.sonarsource.com/products/sonarqube/downloads/lts/).

There will likely be numerous Quarkus users wondering how to deal with that soon, since Quarkus's documentation implicitly suggests field injection, as can be seen in [this prominent example](https://quarkus.io/guides/cdi#what-does-a-bean-look-like). Moreover, SonarQube's check has `defaultSeverity` Major and a impact defined as MEDIUM. Our own projects have roughly a thousand affected fields. [One user was already wondering about this on Stack Overflow nearly a year ago](https://stackoverflow.com/questions/77236634/field-injection-in-quarkus).

It would therefore be quite interesting for Quarkus to take an official stance on this matter, and perhaps provide an opinion about how severe this issue truly is in the context of Quarkus. If the impact is lower as @Ladicek suggests, perhaps SonarQube's check could take that into account. And if the impact truly is medium or higher, then surely Quarkus's documentation should not feature such examples without warnings.

Thank you very much for bringing this up @BrianH85 and all others for your contribution

**Reply by edeandrea** [reply]
HI @Chealer! We're actually in the process of debating this! We're working with the Sonarsource team directly on some proposals.

Feel free to follow/chime in: https://github.com/quarkusio/quarkus/discussions/43179#discussioncomment-10602965

**Reply by Chealer** [reply]
I had completely missed that; thank you so much @edeandrea, that's amazing!
@Ladicek , @manovotn: You might want to chime in there.

**Reply by mkouba** [reply]
I have to admit that I strongly disagree with such warnings - see my comment here: https://github.com/quarkusio/quarkus/discussions/43179#discussioncomment-10828242.

**Comment by omasseau** [comment]
Sonarqube is still reporting this problem when field injection is used.

In every single page of the Quarkus documentation only field injection is used.

I think that either the Quarkus team should : 

- make an official statement in the documentation that field injection is indeed the recommended way to do injection in Quarkus (and inform the Sonarqube devs of this)
- or the documentation should be adapted to use constructor injection if the Quarkus team thinks that constructor injection is still preferred over field injection. 
- or say that neither is preferred over the other one and that both ways are legitimate.

Let's not forget that many junior developers learn the best practices through the documentation of the frameworks they use.

**Reply by omasseau** [reply]
Followup: @see https://github.com/quarkusio/quarkus/issues/48480
